### PR TITLE
Additional task data

### PR DESCRIPTION
**Pull request by kolszewska**
## Proposed Changes

  - Created DB migration adding a `description` and `label_examples` columns to the DB. Additionaly created property `number_of_available_scans` to a Task table.
  - Extended existing endpoint for returning the metadata
  - Created and updated existins tests

## API changes (optional)

When creating a task there is now an option to provide both description and the label_examples (which is just an arrary of strings). The fields are `optional`. The new schema looks like this.

```
{
  "key": "string",
  "name": "string",
  "image_path": "string",
  "datasets_keys": [
    "string"
  ],
  "description": "string",
  "label_examples": [
    "string"
  ],
  "tags": [
    {
      "key": "string",
      "name": "string",
      "actions_ids": [
        0
      ],
      "tools": [
        "string"
      ]
    }
  ]
}
```

Already existing endpoint has been extended - `tasks/{task_key}` which will return the following information about a task (addition of `description`, `label_examples` and `number_of_available_scans`:
```
{
  "key": "string",
  "name": "string",
  "image_path": "string",
  "tags": [
    {
      "key": "string",
      "name": "string",
      "actions_ids": [
        0
      ],
      "tools": [
        "string"
      ]
    }
  ],
  "number_of_available_scans": 0,
  "description": "string",
  "label_examples": [
    "string"
  ],
  "datasets_keys": [
    "string"
  ]
}
```

### CODE REVIEW

File: backend/tests/functional_tests/scripts/test_sync_configuration.py [hunk 1]
```diff
@@ -448,9 +478,7 @@ def test_sync_with_changed_dataset_and_reused_task(prepare_environment: Any) ->
     """)
     script.sync_configuration(configuration)
 
-    # Check if Datasets were synchronized properly
-    datasets = DatasetsRepository.get_all_datasets(include_disabled=True)
```
Comment (jpowie01) [inline backend/tests/functional_tests/scripts/test_sync_configuration.py]: Why did you remove this check? Something has changed or was failing?

File: backend/medtagger/api/tasks/service_rest.py [hunk 2]
```diff
@@ -59,3 +61,18 @@ class Task(Resource):
     def get(task_key: str) -> Any:
         """Return task for given key."""
         return business.get_task_for_key(task_key)
+
+
+@tasks_ns.route('/<string:task_key>/metadata')
```
Comment (jpowie01) [inline backend/medtagger/api/tasks/service_rest.py]: I don't like the idea to have two endpoints for the same business entity. Let's keep only one of them and make it simple.

File: backend/tests/functional_tests/scripts/test_sync_configuration.py [hunk 3]
```diff
@@ -373,6 +397,8 @@ def test_sync_with_changed_task_in_dataset(mocker: Any, prepare_environment: Any
     kidneys_segmentation = TasksRepository.get_task_by_key('KIDNEYS_SEGMENTATION')
     assert kidneys_segmentation.name == 'New Kidneys segmentation'
     assert kidneys_segmentation.image_path == 'assets/icon/kidneys_dataset_icon.svg'
+    assert kidneys_segmentation.description == 'This is a test task'
+    assert len(kidneys_segmentation.label_examples) == 2
```
Comment (jpowie01) [inline backend/tests/functional_tests/scripts/test_sync_configuration.py]: Can we also check what's inside of `kidneys_segmentation.label_examples`? At least in one place ;)

File: backend/tests/functional_tests/scripts/test_sync_configuration.py (hunk 1 above)
Comment (kolszewska) [reply backend/tests/functional_tests/scripts/test_sync_configuration.py]: No, I have no recollection of deleting these lines. It must have been done by mistake :smile:

File: backend/medtagger/database/models.py [hunk 4]
```diff
@@ -164,6 +165,15 @@ def available_tags(self, new_tags: List['LabelTag']) -> None:
         """Set new Label Tags for this Task."""
         self._tags = new_tags
 
+    @property
+    def number_of_available_scans(self) -> int:
+        """Returns a number of available Scans for this Task."""
+        datasets_ids = [dataset.id for dataset in self.datasets]
+        number_of_scans = Scan.query.filter(and_(
+            Scan.dataset_id.in_(datasets_ids)
```
Comment (jpowie01) [inline backend/medtagger/database/models.py]: Please add another filter and return only `ScanStatus.AVAILABLE` Scans.

File: backend/tests/functional_tests/scripts/test_sync_configuration.py [hunk 5]
```diff
@@ -51,6 +53,8 @@ def test_sync_with_empty_database(prepare_environment: Any) -> None:
     kidneys_segmentation = TasksRepository.get_task_by_key('KIDNEYS_SEGMENTATION')
     assert kidneys_segmentation.name == 'Kidneys segmentation'
     assert kidneys_segmentation.image_path == 'assets/icon/kidneys_dataset_icon.svg'
+    assert kidneys_segmentation.description == 'This is a test task'
+    assert kidneys_segmentation.label_examples == ['assets/example_1', 'assets/example_2']
```
Comment (jpowie01) [inline backend/tests/functional_tests/scripts/test_sync_configuration.py]: This is wrong way to compare list. PostgreSQL may return this list in different order.

Safer way to do this:
```python
assert set(kidneys_segmentation.label_examples) == {'assets/example_1', 'assets/example_2'}
```

File: backend/medtagger/api/tasks/business.py [hunk 6]
```diff
@@ -1,4 +1,5 @@
 """Module responsible for business logic in all Tasks endpoints."""
+# pylint: disable-msg=too-many-arguments
```
Comment (jpowie01) [inline backend/medtagger/api/tasks/business.py]: Please do it on the function level if needed. Disabling this check here will have an impact on whole module.

File: backend/medtagger/repositories/tasks.py [hunk 7]
```diff
@@ -1,4 +1,5 @@
 """Module responsible for definition of TaskRepository."""
+# pylint: disable-msg=too-many-arguments
```
Comment (jpowie01) [inline backend/medtagger/repositories/tasks.py]: The same as above.

File: backend/medtagger/database/models.py [hunk 8]
```diff
@@ -164,6 +165,16 @@ def available_tags(self, new_tags: List['LabelTag']) -> None:
         """Set new Label Tags for this Task."""
         self._tags = new_tags
 
+    @property
+    def number_of_available_scans(self) -> int:
```
Comment (jpowie01) [inline backend/medtagger/database/models.py]: Please add also a method `get_number_of_labelled_scans`, which will use this query:

```python
def get_number_of_labeled_scans(self, user_id: int) -> int:
    datasets_ids = [dataset.id for dataset in self.datasets]
    number_of_labeled_scans = Scan.query.join(Label).filter(and_(
        Scan.dataset_id.in_(datasets_ids),
        Scan.status == ScanStatus.AVAILABLE,
        Label.owner_id == user_id
    )).count()
```

Also, please make sure to return this number in the REST API endpoint output as well. Dawid needs it on the frontend side to properly render labeling progress.

Comment (jpowie01) [reply backend/medtagger/database/models.py]: Or use `User` object as an input. Use whatever will be easier for you:

```python
def get_number_of_labeled_scans(self, user: User) -> int:
    ...
```

Comment (jpowie01) [reply backend/medtagger/database/models.py]: After discussion offline, let's do it in another PR.

**Pull request merged**